cartographer/mapping/internal/2d/ray_to_pixel_mask_test.cc

```diff
@@ -0,0 +1,221 @@
+  ray = RayToPixelMask(end, begin, subpixel_scale);
+  EXPECT_EQ(ray.size(), 3);
+  EXPECT_TRUE(ray[0].isApprox(Eigen::Array2i({1, 3})));
+  EXPECT_TRUE(ray[1].isApprox(Eigen::Array2i({2, 2})));
+  EXPECT_TRUE(ray[2].isApprox(Eigen::Array2i({3, 1})));
+}
+
+TEST(RayToPixelMaskTest, SteepLine) {
+  Eigen::Array2i begin = {1, 1};
+  Eigen::Array2i end = {2, 5};
+  const int subpixel_scale = 1;
+  std::vector<Eigen::Array2i> ray = RayToPixelMask(begin, end, subpixel_scale);
+  EXPECT_EQ(ray.size(), 6);
+  EXPECT_TRUE(ray[0].isApprox(Eigen::Array2i({1, 1})));
+  EXPECT_TRUE(ray[1].isApprox(Eigen::Array2i({1, 2})));
+  EXPECT_TRUE(ray[2].isApprox(Eigen::Array2i({1, 3})));
+  EXPECT_TRUE(ray[3].isApprox(Eigen::Array2i({2, 3})));
+  EXPECT_TRUE(ray[4].isApprox(Eigen::Array2i({2, 4})));
+  EXPECT_TRUE(ray[5].isApprox(Eigen::Array2i({2, 5})));
+
+  begin = {1, 1};
+  end = {2, 4};
+  ray = RayToPixelMask(begin, end, subpixel_scale);
+  EXPECT_EQ(ray.size(), 4);
+  EXPECT_TRUE(ray[0].isApprox(Eigen::Array2i({1, 1})));
+  EXPECT_TRUE(ray[1].isApprox(Eigen::Array2i({1, 2})));
+  EXPECT_TRUE(ray[2].isApprox(Eigen::Array2i({2, 3})));
+  EXPECT_TRUE(ray[3].isApprox(Eigen::Array2i({2, 4})));
+}
+
+TEST(RayToPixelMaskTest, FlatLine) {
+  Eigen::Array2i begin = {1, 1};
+  Eigen::Array2i end = {5, 2};
+  const int subpixel_scale = 1;
+  std::vector<Eigen::Array2i> ray = RayToPixelMask(begin, end, subpixel_scale);
+  EXPECT_EQ(ray.size(), 6);
+  EXPECT_TRUE(ray[0].isApprox(Eigen::Array2i({1, 1})));
+  EXPECT_TRUE(ray[1].isApprox(Eigen::Array2i({2, 1})));
+  EXPECT_TRUE(ray[2].isApprox(Eigen::Array2i({3, 1})));
+  EXPECT_TRUE(ray[3].isApprox(Eigen::Array2i({3, 2})));
+  EXPECT_TRUE(ray[4].isApprox(Eigen::Array2i({4, 2})));
+  EXPECT_TRUE(ray[5].isApprox(Eigen::Array2i({5, 2})));
+
+  begin = {1, 1};
+  end = {4, 2};
+  ray = RayToPixelMask(begin, end, subpixel_scale);
+  EXPECT_EQ(ray.size(), 4);
+  EXPECT_TRUE(ray[0].isApprox(Eigen::Array2i({1, 1})));
+  EXPECT_TRUE(ray[1].isApprox(Eigen::Array2i({2, 1})));
+  EXPECT_TRUE(ray[2].isApprox(Eigen::Array2i({3, 2})));
+  EXPECT_TRUE(ray[3].isApprox(Eigen::Array2i({4, 2})));
+}
+
+TEST(RayToPixelMaskTest, MultiScaleAxisAlignedX) {
+  int subpixel_scale = 1;
+  const int num_cells_x = 20;
+  const int num_cells_y = 20;
+  double resolution = 0.1;
+  Eigen::Vector2d max = {1.0, 1.0};
+  double superscaled_resolution = resolution / subpixel_scale;
+  MapLimits superscaled_limits(
+      superscaled_resolution, max,
+      CellLimits(num_cells_x * subpixel_scale, num_cells_y * subpixel_scale));
+  Eigen::Array2i begin =
+      superscaled_limits.GetCellIndex(Eigen::Vector2f({0.05, 0.05}));
+  Eigen::Array2i end =
+      superscaled_limits.GetCellIndex(Eigen::Vector2f({0.35, 0.05}));
+  const std::vector<Eigen::Array2i> base_scale_ray =
+      RayToPixelMask(begin, end, subpixel_scale);
+  EXPECT_EQ(base_scale_ray.size(), 4);
+  EXPECT_TRUE(base_scale_ray[0].isApprox(Eigen::Array2i({9, 6})));
```

Comment:
Couldn't this be in the loop when you start the loop with scale 1?

Reply:
Yes.